src/Symfony/Component/Dsn/DsnParser.php

```diff
@@ -0,0 +1,172 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Symfony\Component\Dsn;
+
+use Symfony\Component\Dsn\Configuration\Dsn;
+use Symfony\Component\Dsn\Configuration\DsnFunction;
+use Symfony\Component\Dsn\Configuration\Path;
+use Symfony\Component\Dsn\Configuration\Url;
+use Symfony\Component\Dsn\Exception\FunctionsNotAllowedException;
+use Symfony\Component\Dsn\Exception\SyntaxException;
+
+/**
+ * A factory class to parse a string and create a DsnFunction.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class DsnParser
+{
+    private const FUNCTION_REGEX = '#^([a-zA-Z0-9\+-]+):?\((.*)\)(?:\?(.*))?$#';
+    private const ARGUMENTS_REGEX = '#([^\s,]+\(.+\)(?:\?.*)?|[^\s,]+)#';
+    private const UNRESERVED = 'a-zA-Z0-9-\._~';
+    private const SUB_DELIMS = '!\$&\'\(\}\*\+,;=';
+
+    /**
+     * Parse A DSN thay may contain functions. If no function is present in the
+     * string, then a "dsn()" function will be added.
+     *
+     * @throws SyntaxException
+     */
+    public static function parseFunc(string $dsn): DsnFunction
```

Comment:
shouldn't that return `Dns|DnsFunction` instead of a fake function named `dsn` ?

Reply:
That would force the user to always check what object they got back. 

```php
$dsn = DsnParser::parseFunc($userInput);
if ($dsn instanceof DsnFunction) {
  // Handle function logic
  // Foreach (arguments as argument)
      // handle single argument
} else {
  // Handle single argument
}
```

But with constant return types: 
```php
$dsn = DsnParser::parseFunc($userInput);
  // Foreach (arguments as argument)
      // handle single argument
```